src/prelude/internals.mo

```diff
@@ -588,31 +588,34 @@ func @timer_helper() : async () {
   func gatherExpired(n : ?@Node) = switch n {
     case null ();
     case (?n) {
-      gatherExpired(n.pre);
-      if (n.expire[0] > 0 and n.expire[0] <= now and gathered < thunks.size()) {
-        thunks[gathered] := ?(n.job);
-        switch (n.delay) {
-          case (null or ?0) ();
-          case (?delay) {
-            // re-add the node, skipping past expirations
-            let expire = n.expire[0] + delay * (1 + (now - n.expire[0]) / delay);
-            n.expire[0] := 0;
-            // N.B. reinsert only works on pruned nodes
-            func reinsert(m : ?@Node) : @Node = switch m {
-              case null ({ n with expire = [var expire]; pre = null; post = null });
-              case (?m) {
-                assert m.expire[0] != 0;
-                if (expire < m.expire[0]) ({ m with pre = ?reinsert(m.pre) })
-                else ({ m with post = ?reinsert(m.post) })
-              }
-            };
-            @timers := ?reinsert(@prune(@timers));
+      let pivot = n.expire[0];
+      if (pivot > 0) gatherExpired(n.pre); // by corollary
+      if (pivot <= now and gathered < thunks.size()) {
+        if (pivot > 0) {
+          // not expunged yet
+          thunks[gathered] := ?(n.job);
+          switch (n.delay) {
+            case (null or ?0) n.expire[0] := 0;
+            case (?delay) {
+              // re-add the node, skipping past expirations
+              let expire = pivot + delay * (1 + (now - pivot) / delay);
+              n.expire[0] := 0;
+              // N.B. reinsert only works on pruned nodes
+              func reinsert(m : ?@Node) : @Node = switch m {
+                case null ({ n with expire = [var expire]; pre = null; post = null });
+                case (?m) {
+                  assert m.expire[0] != 0;
+                  if (expire < m.expire[0]) ({ m with pre = ?reinsert(m.pre) })
+                  else ({ m with post = ?reinsert(m.post) })
+                }
+              };
+              @timers := ?reinsert(@prune(@timers));
```

Comment:
This does not have any effect on the ongoing `gatherExpired`, right? `reinsert` is not mutating anything.
So could we avoid calling `prune` and `reinsert` for every gathered job and simply keep nodes that need to be re-added later in a max 10-element array and add them after `gatherExpired` finished? After sorting the array it should be possible to traverse `timers` just once.
Note: this is assuming that re-inserted nodes are not to be gathered immediately.

Reply:
Right, the current traversal is not touched by reinsertion. And yes we could optimise more, but let's do it in small increments!

Reply:
Also, I expect a very small number of recurrent timers to expire at any point (< 2, I guess). So the benefit is not sure.